R/gm-auth.R

```diff
@@ -0,0 +1,207 @@
+# This file is the interface between gmailr and the
+# auth functionality in gargle.
+.auth <- gargle::init_AuthState(
+  package     = "gmailr",
+  auth_active = TRUE # TODO: what is this?
+)
+
+# The roxygen comments for these functions are mostly generated from data
+# in this list and template text maintained in gargle.
+gargle_lookup_table <- list(
+  PACKAGE     = "gmailr",
+  YOUR_STUFF  = "your Gmail projects",
+  PRODUCT     = "Google Gmail",
+  API         = "Gmail API",
+  PREFIX      = "gm",
+  AUTH_CONFIG_SOURCE = "gmailr" #TODO: what is this?
+)
+
+#' Authorize bigrquery
+#'
+#' @eval gargle:::PREFIX_auth_description(gargle_lookup_table)
+#' @eval gargle:::PREFIX_auth_details(gargle_lookup_table)
+#' @eval gargle:::PREFIX_auth_params()
+#'
+#' @family auth functions
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' ## load/refresh existing credentials, if available
+#' ## otherwise, go to browser for authentication and authorization
+#' gm_auth()
+#'
+#' ## force use of a token associated with a specific email
+#' gm_auth(email = "jim@example.com")
+#'
+#' ## force a menu where you can choose from existing tokens or
+#' ## choose to get a new one
+#' gm_auth(email = NA)
+#'
+#' ## use a 'read only' scope, so it's impossible to change data
+#' gm_auth(
+#'   scopes = "https://www.googleapis.com/auth/gmail.readonly"
+#' )
+#'
+#' ## use a service account token
+#' gm_auth(path = "foofy-83ee9e7c9c48.json")
+#' }
+gm_auth <- function(email = NULL,
+                    path = NULL,
+                    scopes = c(
+                      "https://www.googleapis.com/auth/gmail.readonly",
+                      "https://www.googleapis.com/auth/gmail.modify",
+                      "https://www.googleapis.com/auth/gmail.compose",
+                      "https://mail.google.com/"
+                    ),
+                    cache = gargle::gargle_oauth_cache(),
+                    use_oob = gargle::gargle_oob_default(),
+                    token = NULL) {
+  cred <- gargle::token_fetch(
+    scopes = scopes,
+    app = gm_oauth_app(),
+    email = email,
+    path = path,
+    package = "gmailr",
+    cache = cache,
+    use_oob = use_oob,
+    token = token
+  )
+  if (!inherits(cred, "Token2.0")) {
+    stop(
+      "Can't get Google credentials.\n",
+      "Are you running gmailr in a non-interactive session? Consider:\n",
+      "  * Call `gm_auth()` directly with all necessary specifics.\n",
+      call. = FALSE
+    )
+  }
+  .auth$set_cred(cred)
+  .auth$set_auth_active(TRUE)
+
+  invisible()
+}
+
+#' Clear current token
+#'
+#' Clears any currently stored token. The next time gmailr needs a token, the
```

Comment:
There is now a template for this too:

https://github.com/r-lib/gargle/blob/1870dc19475eb8f68d454b8a9db4c27c72be37e6/R/roxygen-templates.R#L99

It got added to gargle after bigrquery went to CRAN. But you can use it.